### PR TITLE
feat(Variants): Add Arduino libraries for Lora 4G Module and Node Wifi o…

### PR DESCRIPTION
**Pull request by Bharat-Pi**
…f BharatPi boards variant.

## Description of Change
This Pull Request add support to three new hardware on the market using ESP32 modules, Bharat Pi Node Wifi , Bharat Pi A7672s 4G and Bharat Pi LoRa modules. Links to the hardware is provided below

Bharat Pi Node Wifi
https://bharatpi.net/product/bharat-pi-node-wifi/

Bharat Pi A7672s 4G Board
https://bharatpi.net/product/bharat-pi-4g-lte-module-with-esp32-simcom-a7672-with-gps/

Bharat Pi LoRa boards with Indian Frequency
https://bharatpi.net/product/bharat-pi-lora-module-with-esp32-lora868mhz-transreceiver-with-esp32-arduino-compatible/

These boards are used by Students, Tinkerers across 100s of different colleges and universities and we are connected with the Espressif to launch these boards.

## Tests scenarios
We have tested these changes on a Bharat Pi boards using the Arduino IDE and the ESP32 core with 4MB, 8MB and 16MB flash versions of ESP32.
We have verified that the defined pins correspond to the correct hardware pins on the board.
We tested the samples codes and libraries of our code and example codes.
You can see our test codes in this page
https://github.com/Bharat-Pi/iot-projects

## Related links
This is a support package for new variant
https://github.com/Bharat-Pi/iot-projects
https://bharatpi.net/product/bharat-pi-node-wifi/
https://bharatpi.net/product/bharat-pi-4g-lte-module-with-esp32-simcom-a7672-with-gps/
https://bharatpi.net/product/bharat-pi-lora-module-with-esp32-lora868mhz-transreceiver-with-esp32-arduino-compatible/

### CODE REVIEW

File: variants/BharatPi-Node-Wifi/pins_arduino.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+static const uint8_t A3 = 4;
+static const uint8_t A4 = 2;
+static const uint8_t A5 = 0;
+
+static const uint8_t TX = 1;
+static const uint8_t RX = 3;
+
+static const uint8_t TX2 = 17;
+static const uint8_t RX2 = 16;
+#endif /* Pins_Arduino_h */
```
Comment (P-R-O-C-H-Y) [inline variants/BharatPi-Node-Wifi/pins_arduino.h:51]: This part is doubled, please remote it.
```suggestion
#endif /* Pins_Arduino_h */
```

File: variants/BharatPi-A7672S-4G/pins_arduino.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#ifndef Pins_Arduino_h
+#define Pins_Arduino_h
+
+#include <stdint.h>
+
+static const uint8_t LED_BUILTIN = 2;
+#define BUILTIN_LED LED_BUILTIN  // backward compatibility
+#define LED_BUILTIN LED_BUILTIN  // allow testing #ifdef LED_BUILTIN
+
```
Comment (P-R-O-C-H-Y) [inline variants/BharatPi-A7672S-4G/pins_arduino.h:9]: You are missing default SPI and I2C pins, please add them.
This must be done for all boards.
If you don't want to specify the default pin, you can use -1 as the pin so defaults from the core will be used.

**Review by P-R-O-C-H-Y — CHANGES_REQUESTED** [review]

File: boards.txt [hunk 3]
```diff
@@ -2800,6 +2800,479 @@ aventen_s3_sync.menu.EraseFlash.all.upload.erase_cmd=-e
...
+BharatPi-Node-Wifi.serial.disableDTR=true
+BharatPi-Node-Wifi.serial.disableRTS=true
+
+BharatPi-Node-Wifi.build.tarch=xtensa
+BharatPi-Node-Wifi.build.bootloader_addr=0x1000
+BharatPi-Node-Wifi.build.target=esp32
+BharatPi-Node-Wifi.build.mcu=esp32
+BharatPi-Node-Wifi.build.core=esp32
+BharatPi-Node-Wifi.build.variant=BharatPi-Node-Wifi
+BharatPi-Node-Wifi.build.board=ESP32_WROOM_DA
```
Comment (P-R-O-C-H-Y) [inline boards.txt:2826]: Please change this to proper name.
```suggestion
BharatPi-Node-Wifi.build.board=BHARATPI_NODE_WIFI
```

File: boards.txt [hunk 4]
```diff
@@ -2800,6 +2800,479 @@ aventen_s3_sync.menu.EraseFlash.all.upload.erase_cmd=-e
...
+BharatPi-Node-Wifi.menu.FlashFreq.40=40MHz
+BharatPi-Node-Wifi.menu.FlashFreq.40.build.flash_freq=40m
+
+BharatPi-Node-Wifi.menu.FlashSize.4M=4MB (32Mb)
+BharatPi-Node-Wifi.menu.FlashSize.4M.build.flash_size=4MB
+BharatPi-Node-Wifi.menu.FlashSize.8M=8MB (64Mb)
+BharatPi-Node-Wifi.menu.FlashSize.8M.build.flash_size=8MB
+BharatPi-Node-Wifi.menu.FlashSize.8M.build.partitions=default_8MB
+BharatPi-Node-Wifi.menu.FlashSize.16M=16MB (128Mb)
+BharatPi-Node-Wifi.menu.FlashSize.16M.build.flash_size=16MB
```
Comment (P-R-O-C-H-Y) [inline boards.txt:2913]: You can remove those as the board is made with 4MB flash chip
```suggestion
```

File: boards.txt [hunk 5]
```diff
@@ -2800,6 +2800,479 @@ aventen_s3_sync.menu.EraseFlash.all.upload.erase_cmd=-e
...
+BharatPi-Node-Wifi.build.flash_size=4MB
+BharatPi-Node-Wifi.build.flash_freq=40m
+BharatPi-Node-Wifi.build.flash_mode=dio
+BharatPi-Node-Wifi.build.boot=dio
+BharatPi-Node-Wifi.build.partitions=default
+BharatPi-Node-Wifi.build.defines=
+BharatPi-Node-Wifi.build.loop_core=
+BharatPi-Node-Wifi.build.event_core=
+
+BharatPi-Node-Wifi.menu.PartitionScheme.default=Default 4MB with spiffs (1.2MB APP/1.5MB SPIFFS)
```
Comment (P-R-O-C-H-Y) [inline boards.txt:2838]: Also go through all partition schemes and remove those that are not for 4MB Flash size.

File: boards.txt [hunk 6]
```diff
@@ -2800,6 +2800,479 @@ aventen_s3_sync.menu.EraseFlash.all.upload.erase_cmd=-e
...
+BharatPi-A7672S-4G.serial.disableDTR=true
+BharatPi-A7672S-4G.serial.disableRTS=true
+
+BharatPi-A7672S-4G.build.tarch=xtensa
+BharatPi-A7672S-4G.build.bootloader_addr=0x1000
+BharatPi-A7672S-4G.build.target=esp32
+BharatPi-A7672S-4G.build.mcu=esp32
+BharatPi-A7672S-4G.build.core=esp32
+BharatPi-A7672S-4G.build.variant=BharatPi-A7672S-4G
+BharatPi-A7672S-4G.build.board=ESP32_WROOM_DA
```
Comment (P-R-O-C-H-Y) [inline boards.txt:2984]: Same here:
```suggestion
BharatPi-A7672S-4G.build.board=BHARATPI_A7672S_4G
```

File: boards.txt [hunk 7]
```diff
@@ -2800,6 +2800,479 @@ aventen_s3_sync.menu.EraseFlash.all.upload.erase_cmd=-e
...
+BharatPi-A7672S-4G.menu.FlashFreq.40=40MHz
+BharatPi-A7672S-4G.menu.FlashFreq.40.build.flash_freq=40m
+
+BharatPi-A7672S-4G.menu.FlashSize.4M=4MB (32Mb)
+BharatPi-A7672S-4G.menu.FlashSize.4M.build.flash_size=4MB
+BharatPi-A7672S-4G.menu.FlashSize.8M=8MB (64Mb)
+BharatPi-A7672S-4G.menu.FlashSize.8M.build.flash_size=8MB
+BharatPi-A7672S-4G.menu.FlashSize.8M.build.partitions=default_8MB
+BharatPi-A7672S-4G.menu.FlashSize.16M=16MB (128Mb)
+BharatPi-A7672S-4G.menu.FlashSize.16M.build.flash_size=16MB
```
Comment (P-R-O-C-H-Y) [inline boards.txt:3071]: Same here:
```suggestion
```

File: boards.txt [hunk 8]
```diff
@@ -2800,6 +2800,479 @@ aventen_s3_sync.menu.EraseFlash.all.upload.erase_cmd=-e
...
+BharatPi-LoRa.serial.disableDTR=true
+BharatPi-LoRa.serial.disableRTS=true
+
+BharatPi-LoRa.build.tarch=xtensa
+BharatPi-LoRa.build.bootloader_addr=0x1000
+BharatPi-LoRa.build.target=esp32
+BharatPi-LoRa.build.mcu=esp32
+BharatPi-LoRa.build.core=esp32
+BharatPi-LoRa.build.variant=BharatPi-LoRa
+BharatPi-LoRa.build.board=ESP32_WROOM_DA
```
Comment (P-R-O-C-H-Y) [inline boards.txt:3142]: Same here:
```suggestion
BharatPi-LoRa.build.board=BHARATPI_LORA
```

File: boards.txt [hunk 9]
```diff
@@ -2800,6 +2800,479 @@ aventen_s3_sync.menu.EraseFlash.all.upload.erase_cmd=-e
...
+BharatPi-LoRa.menu.FlashFreq.40=40MHz
+BharatPi-LoRa.menu.FlashFreq.40.build.flash_freq=40m
+
+BharatPi-LoRa.menu.FlashSize.4M=4MB (32Mb)
+BharatPi-LoRa.menu.FlashSize.4M.build.flash_size=4MB
+BharatPi-LoRa.menu.FlashSize.8M=8MB (64Mb)
+BharatPi-LoRa.menu.FlashSize.8M.build.flash_size=8MB
+BharatPi-LoRa.menu.FlashSize.8M.build.partitions=default_8MB
+BharatPi-LoRa.menu.FlashSize.16M=16MB (128Mb)
+BharatPi-LoRa.menu.FlashSize.16M.build.flash_size=16MB
```
Comment (P-R-O-C-H-Y) [inline boards.txt:3229]: If it also uses 4MB flash please remove those options and remove the unusable partitions.

**Review by P-R-O-C-H-Y — CHANGES_REQUESTED** [review]
@Bharat-Pi PTAL on my comments.
As you are using 4MB Flash chip, please remove other options and unusable partitions for all the boards. Thanks, otherwise its looking good.

**Comment by VojtechBartoska** [comment]
Hello @Bharat-Pi, please take a look on proposed changes by @P-R-O-C-H-Y.

File: boards.txt [hunk 10]
```diff
@@ -2780,6 +2780,479 @@ aventen_s3_sync.menu.EraseFlash.all.upload.erase_cmd=-e
...
+BharatPi-Node-Wifi.menu.FlashFreq.40=40MHz
+BharatPi-Node-Wifi.menu.FlashFreq.40.build.flash_freq=40m
+
+BharatPi-Node-Wifi.menu.FlashSize.4M=4MB (32Mb)
+BharatPi-Node-Wifi.menu.FlashSize.4M.build.flash_size=4MB
+BharatPi-Node-Wifi.menu.FlashSize.8M=8MB (64Mb)
+BharatPi-Node-Wifi.menu.FlashSize.8M.build.flash_size=8MB
+BharatPi-Node-Wifi.menu.FlashSize.8M.build.partitions=default_8MB
+BharatPi-Node-Wifi.menu.FlashSize.16M=16MB (128Mb)
+BharatPi-Node-Wifi.menu.FlashSize.16M.build.flash_size=16MB
```
Comment (P-R-O-C-H-Y) [inline boards.txt:2893]: Please remove this as it's not possible option for your boards. Apply to all of them.
```suggestion
```

**Comment by Bharat-Pi** [comment]
> Hello @Bharat-Pi, please take a look on proposed changes by @P-R-O-C-H-Y.

Somehow my comments are not shown 
Here is the comments I mentioned 

Though the board picture you see in the website has 4MB flash, We have boards with 8MB and 16MB Flash versions, which are used by ones who wants more flash memory and trying edge ai workloads. I have also mentioned the same in the description.

This is same for all three boards, Hence we would need partition and flash size of 8MB and 16MB also. 

It is showing as pending for me not sure why . I have resolved other comments regarding board name , please check and approve the PR
image

File: boards.txt [hunk 11]
```diff
@@ -2780,6 +2780,479 @@ aventen_s3_sync.menu.EraseFlash.all.upload.erase_cmd=-e
...
+BharatPi-Node-Wifi.menu.PartitionScheme.huge_app.upload.maximum_size=3145728
+BharatPi-Node-Wifi.menu.PartitionScheme.min_spiffs=Minimal SPIFFS (1.9MB APP with OTA/190KB SPIFFS)
+BharatPi-Node-Wifi.menu.PartitionScheme.min_spiffs.build.partitions=min_spiffs
+BharatPi-Node-Wifi.menu.PartitionScheme.min_spiffs.upload.maximum_size=1966080
+BharatPi-Node-Wifi.menu.PartitionScheme.fatflash=16M Flash (2MB APP/12.5MB FATFS)
+BharatPi-Node-Wifi.menu.PartitionScheme.fatflash.build.partitions=ffat
+BharatPi-Node-Wifi.menu.PartitionScheme.fatflash.upload.maximum_size=2097152
+BharatPi-Node-Wifi.menu.PartitionScheme.app3M_fat9M_16MB=16M Flash (3MB APP/9.9MB FATFS)
+BharatPi-Node-Wifi.menu.PartitionScheme.app3M_fat9M_16MB.build.partitions=app3M_fat9M_16MB
+BharatPi-Node-Wifi.menu.PartitionScheme.app3M_fat9M_16MB.upload.maximum_size=3145728
```
Comment (P-R-O-C-H-Y) [inline boards.txt:2850]: Also remove those partitions -> needs 16MB of Flash. Apply to all boards please.
```suggestion
```

**Review by P-R-O-C-H-Y — CHANGES_REQUESTED** [review]
@Bharat-Pi PTAL

File: boards.txt [hunk 12]
```diff
@@ -2780,6 +2780,479 @@ aventen_s3_sync.menu.EraseFlash.all.upload.erase_cmd=-e
...
+BharatPi-A7672S-4G.serial.disableDTR=true
+BharatPi-A7672S-4G.serial.disableRTS=true
+
+BharatPi-A7672S-4G.build.tarch=xtensa
+BharatPi-A7672S-4G.build.bootloader_addr=0x1000
+BharatPi-A7672S-4G.build.target=esp32
+BharatPi-A7672S-4G.build.mcu=esp32
+BharatPi-A7672S-4G.build.core=esp32
+BharatPi-A7672S-4G.build.variant=BharatPi-A7672S-4G
+BharatPi-A7672S-4G.build.board=BHARATPI_A7672s_4G
```
Comment (P-R-O-C-H-Y) [inline boards.txt:2964]: The build.board must be uppercase, you missed the "s". Please fix it.
```suggestion
BharatPi-A7672S-4G.build.board=BHARATPI_A7672S_4G
```

**Comment by P-R-O-C-H-Y** [comment]
@Bharat-Pi Thanks for the info, one small nitpick about the build.board name. Else it's looking good. Once fixed and CI passes we can merge :)

**Comment by Bharat-Pi** [comment]
> @Bharat-Pi Thanks for the info, one small nitpick about the build.board name. Else it's looking good. Once fixed and CI passes we can merge :)

Fixed and pushed the changes, Thanks for the quick support. Hope it gets merged today

**Review by P-R-O-C-H-Y — APPROVED** [review]

**Pull request merged**